NRE in build task

"E:\git\temp\temp.csproj" (default target) (1) ->
(CoreCompile target) ->
  E:\git\cli-1\artifacts\win10-x64\stage2\sdk\1.0.0-preview3-003633\Microsoft.CSharp.Core.targets(71,5): error MSB3883: Unexpected exception:  [E:\git\temp\temp.csproj]
  E:\git\cli-1\artifacts\win10-x64\stage2\sdk\1.0.0-preview3-003633\Microsoft.CSharp.Core.targets(71,5): error : System.NullReferenceException: Object reference not set to an instance of an object. [E:\git\temp\temp.csproj]
  E:\git\cli-1\artifacts\win10-x64\stage2\sdk\1.0.0-preview3-003633\Microsoft.CSharp.Core.targets(71,5): error : at Microsoft.CodeAnalysis.CommandLine.DesktopBuildClient.GetUserName() [E:\git\temp\temp.csproj]
  E:\git\cli-1\artifacts\win10-x64\stage2\sdk\1.0.0-preview3-003633\Microsoft.CSharp.Core.targets(71,5): error : at Microsoft.CodeAnalysis.CommandLine.DesktopBuildClient.GetPipeNameForPath(String compilerExeDirectory) [E:\git\temp\temp.csproj]
  E:\git\cli-1\artifacts\win10-x64\stage2\sdk\1.0.0-preview3-003633\Microsoft.CSharp.Core.targets(71,5): error : at Microsoft.CodeAnalysis.CommandLine.DesktopBuildClient.RunServerCompilation(RequestLanguage language, List`1 arguments, BuildPaths buildPaths, String keepAlive, String libEnvVariable, CancellationToken cancellationToken) [E:\git\temp\temp.csproj]
  E:\git\cli-1\artifacts\win10-x64\stage2\sdk\1.0.0-preview3-003633\Microsoft.CSharp.Core.targets(71,5): error : at Microsoft.CodeAnalysis.BuildTasks.ManagedCompiler.ExecuteTool(String pathToTool, String responseFileCommands, String commandLineCommands) [E:\git\temp\temp.csproj]

--- a/src/Compilers/Core/MSBuildTask/Utilities.cs
+++ b/src/Compilers/Core/MSBuildTask/Utilities.cs
@@ -20,7 +20,9 @@ namespace Microsoft.CodeAnalysis.BuildTasks
         /// </summary>
         internal static bool IsCompilerServerSupported =>
             RuntimeInformation.IsOSPlatform(OSPlatform.Windows) &&
-            DesktopBuildClient.GetRuntimeDirectoryOpt() != null;
+            DesktopBuildClient.GetRuntimeDirectoryOpt() != null &&
+            // Test that we can retrieve a valid pipe name
+            DesktopBuildClient.GetPipeNameForPathOpt("") != null;
 
         /// <summary>
         /// Convert a task item metadata to bool. Throw an exception if the string is badly formed and can't

--- a/src/Compilers/Server/VBCSCompiler/DesktopBuildServerController.cs
+++ b/src/Compilers/Server/VBCSCompiler/DesktopBuildServerController.cs
@@ -80,7 +80,7 @@ namespace Microsoft.CodeAnalysis.CompilerServer
         protected override string GetDefaultPipeName()
         {
             var clientDirectory = AppDomain.CurrentDomain.BaseDirectory;
-            return DesktopBuildClient.GetPipeNameForPath(clientDirectory);
+            return DesktopBuildClient.GetPipeNameForPathOpt(clientDirectory);
         }
 
         protected override bool? WasServerRunning(string pipeName)

--- a/src/Compilers/Shared/DesktopBuildClient.cs
+++ b/src/Compilers/Shared/DesktopBuildClient.cs
@@ -92,11 +92,18 @@ namespace Microsoft.CodeAnalysis.CommandLine
             string libEnvVariable,
             CancellationToken cancellationToken)
         {
+            var pipeNameOpt = GetPipeNameForPathOpt(buildPaths.ClientDirectory);
+
+            if (pipeNameOpt == null)
+            {
+                return Task.FromResult<BuildResponse>(new RejectedBuildResponse());
+            }
+
             return RunServerCompilationCore(
                 language,
                 arguments,
                 buildPaths,
-                GetPipeNameForPath(buildPaths.ClientDirectory),
+                pipeNameOpt,
                 keepAlive,
                 libEnvVariable,
                 timeoutOverride: null,
@@ -473,13 +480,13 @@ namespace Microsoft.CodeAnalysis.CommandLine
             (ObjectSecurity)typeof(PipeStream)
             .GetTypeInfo()
             .GetDeclaredMethod("GetAccessControl")
-            .Invoke(pipeStream, parameters: null);
+            ?.Invoke(pipeStream, parameters: null);
 
         private static string GetUserName() =>
             (string)typeof(Environment)
             .GetTypeInfo()
             .GetDeclaredProperty("UserName")
-            .GetMethod.Invoke(null, parameters: null);
+            ?.GetMethod?.Invoke(null, parameters: null);
 
         /// <summary>
         /// Given the full path to the directory containing the compiler exes,
@@ -488,10 +495,13 @@ namespace Microsoft.CodeAnalysis.CommandLine
         /// </summary>
         protected override string GetSessionKey(BuildPaths buildPaths)
         {
-            return GetPipeNameForPath(buildPaths.ClientDirectory);
+            return GetPipeNameForPathOpt(buildPaths.ClientDirectory);
         }
 
-        internal static string GetPipeNameForPath(string compilerExeDirectory)
+        /// <returns>
+        /// Null if not enough information was found to create a valid pipe name.
+        /// </returns>
+        internal static string GetPipeNameForPathOpt(string compilerExeDirectory)
         { 
             var basePipeName = GetBasePipeName(compilerExeDirectory);
 
@@ -500,6 +510,11 @@ namespace Microsoft.CodeAnalysis.CommandLine
             var principal = new WindowsPrincipal(currentIdentity);
             var isAdmin = principal.IsInRole(WindowsBuiltInRole.Administrator);
             var userName = GetUserName();
+            if (userName == null)
+            {
+                return null;
+            }
+
             return $"{userName}.{isAdmin}.{basePipeName}";
         }
 

